rakelib/plugins_docs_dependencies.rake

```diff
@@ -203,6 +204,29 @@ task :generate_plugins_version do
         end
       end
     end
+    DepProxy.class_eval do
+      # Bundler caches it's dep-proxy objects (which contain Gem::Dependency objects) from all resolutions.
+      # The Hash itself continues to grow between dependency resolutions and hold up a lot of memory, to avoid
+      # the issue we expose a way of clear-ing the cached objects before each plugin resolution.
+      def self.__clear!
+        @proxies.clear
+      end
+    end
+
+    Fetcher::CompactIndex.class_eval do
+      alias_method :__bundle_worker, :bundle_worker
+      # The compact index is built using multiple threads and this is hard-coded atm to 25 threads:
+      #   `Bundler::Worker.new(Bundler.current_ruby.rbx? ? 1 : 25, worker_name, func)`
+      # each thread might built up a Bundler::Source::Rubygems object which retains more than 100MB.
+      # By limiting the worker thread count we make sure not to produce too many of these objects.
```

Comment:
This is a really good find.

I'm a bit wary of the `Bundler::Fetcher::CompactIndex#bundle_worker` implementation changing out of under us -- if it were to perform an operation that had a side-effect of starting the worker threads, our interception here would be too late to make a difference.

An alternate approach would be to patch `Bundler::Worker` to cap the size of the pool, something like:

~~~ ruby
Worker.class_eval do
  alias_method :__initialize, :initialize
  def initialize(size, name, func)
    if size > 2
      $stderr.puts("Bundler::Worker: initializing limited worker pool (size: 2; requested: #{size})"
      size = 2
    end
    __initialize(size, name, func)
  end
end
~~~

Reply:
Thanks, I understand the concern and have thought about patching the `Worker` as well.
The worker was introduced 4.5+ years ago with [Bundler 2.2](https://github.com/rubygems/rubygems/commit/af1038dea97e513d14af07d6ff430c65a3ab3d07).
(has bee used previously but existed inline - there wasn't a `bundle_worker` method here).

Would say the main use-case for `Bundler::Worker` is the parallel installation ([from the installer](https://github.com/rubygems/rubygems/blob/bundler-v2.2.15/bundler/lib/bundler/installer/parallel_installer.rb#L175-L179)), this can be disabled using the API if we choose to (command line `--jobs` option).

That being said I would rather not cause future surprises by affecting `bundle install`, unless we know it to be problematic. My general approach to patches tends to be to be contained and have minimal effect on other places.

Interestingly the `25` constant seems to have a life of its own, will try to raise this upstream with a reproducer.